assembly/Quat.ts

```diff
@@ -0,0 +1,388 @@
+import {Vec3} from "./Vec3";
+import {Mat4} from "./Mat4";
+import {pc_math} from "./Math";
+
+export class Quat {
+    x: f32;
+    y: f32;
+    z: f32;
+    w: f32;
+
+    constructor(x: f32, y: f32, z: f32, w: f32) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        this.w = w;
+    }
+
+    clone(): Quat {
+        return new Quat(this.x, this.y, this.z, this.w);
+    }
+
+    conjugate(): Quat {
+        this.x *= -1;
+        this.y *= -1;
+        this.z *= -1;
+        return this;
+    }
+
+    copy(rhs: Quat): Quat {
+        this.x = rhs.x;
+        this.y = rhs.y;
+        this.z = rhs.z;
+        this.w = rhs.w;
+        return this;
+    }
+
+    equals(rhs: Quat): boolean {
+        return ((this.x === rhs.x) && (this.y === rhs.y) && (this.z === rhs.z) && (this.w === rhs.w));
+    }
+
+    getAxisAngle(axis: Vec3): f32 {
+        var rad = Mathf.acos(this.w) * 2;
+        var s = Mathf.sin(rad / 2);
+        if (s !== 0) {
+            axis.x = this.x / s;
+            axis.y = this.y / s;
+            axis.z = this.z / s;
+            if (axis.x < 0 || axis.y < 0 || axis.z < 0) {
+                // Flip the sign
+                axis.x *= -1;
+                axis.y *= -1;
+                axis.z *= -1;
+                rad *= -1;
+            }
+        } else {
+            // If s is zero, return any axis (no rotation - axis does not matter)
+            axis.x = 1;
+            axis.y = 0;
+            axis.z = 0;
+        }
+        return rad * pc_math.RAD_TO_DEG;
```

Comment:
```suggestion
        let ax: f32 = 1, ay: f32 = 0, az: f32 = 0;
        if (s !== 0) {
            s = 1.0 / s;
            ax = this.x * s;
            ay = this.y * s;
            az = this.z * s;
            if (ax < 0 || ay < 0 || az < 0) {
                // Flip the sign
                ax *= -1;
                ay *= -1;
                az *= -1;
                rad *= -1;
            }
        }
        axis.x = ax;
        axis.y = ay;
        axis.z = az;
        return rad * pc_math.RAD_TO_DEG;
```

Reply:
Thanks! I understand to remove divisions with faster multiplications, but is there a point of declaring ax/ay/az besides "coding style taste"? Otherwise I would like the code to look nearly exactly like the original JS code

Reply:
It also reduced memory accessing to instance fields